refactor: qb.engine

netchampfaris opened the pull request:
`frappe.qb.engine.get_query` is supposed to be the spiritual successor to `DatabaseQuery.execute` which in turn is used in a lot of places including `frappe.db.get_list` and `frappe.db.get_all`. However, there are a few missing features that are preventing feature parity. When I jumped into `query.py`, I could not get my head around how `qb.engine` worked, after spending a day I figured out that there is a lot of repeated code and unnecessary passing of values and incorrect function names. In short, it needed a refactor. Here is my attempt to refactor the code. Hopefully, this is better to read and extend.

~~All of the tests are passing except the nested set test where results order is different. Will try to fix it too.~~
~~It passes now. The earlier implementation set `desc` as the default order direction. It should be `asc` as per SQL standards.~~

A lot of Frappe code depends on the default order direction to be `desc`. Unless every calling method explicitly passes the direction, it has to be `desc` for now.

I have also changed all instances `qb.engine.get_query` to `qb.get_query` because I ran into a problem where the same instance of `Engine` was modified during a query.

Also, `qb.engine.build_conditions` is not needed, `qb.get_query` should work whether fields or filters are passed or not.

## Features added

### Dynamic fields in filters. Works with link fields and child tables.
```py
                                                         (here)
frappe.qb.get_query("DocType", fields=["name"], filters={"module.app_name": "frappe"})
# SELECT `tabDocType`.`name` FROM `tabDocType` LEFT JOIN `tabModule Def` ON `tabModule Def`.`name`=`tabDocType`.`module` WHERE `tabModule Def`.`app_name`='frappe'

                                                         (here)
frappe.qb.get_query("DocType", fields=["name"], filters={"fields.fieldname": "title"})
# SELECT `tabDocType`.`name` FROM `tabDocType` LEFT JOIN `tabDocField` ON `tabDocField`.`parent`=`tabDocType`.`name` AND `tabDocField`.`parenttype`='DocType' WHERE `tabDocField`.`fieldname`='title'
```

## Perf check

**frappe.qb.engine.get_query** (Before)
```py
In [14]: %timeit frappe.qb.engine.get_query("DocType", fields=["name", "Count(*)", "abs(field1-field2)", "module.app as app"], filters=[["DocField", "name", "like", "f%"],[
    ...: "DocType", "parent", "=", "something"]])
1.34 ms ± 50.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**frappe.qb.get_query** (After)
```py
In [1]: %timeit frappe.qb.get_query("DocType", fields=["name", "Count(*)", "abs(field1-field2)", "module.app as app"], filters=[["DocField", "name", "like", "f%"],["DocType", "pa
   ...: rent", "=", "something"]])
1.19 ms ± 107 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The results are almost identical over several runs. Even if the refactor made it a bit slower, I would prefer it because it adds more features and the code is readable and easily extendable.

netchampfaris committed:
refactor: qb.engine

- simplify
- qb.engine.get_query -> qb.get_query
- qb.engine.build_conditions -> qb.get_query

netchampfaris committed:
fix: delete option

netchampfaris committed:
fix: use table.field instead Field('field')

netchampfaris committed:
fix: refactor usage

netchampfaris committed:
fix: explicitly specifiy order

netchampfaris committed:
fix: allow dynamic fields in filters

e.g.,
`filters={'link.field': 'value'}`
`filters={'child.field': 'value'}`

netchampfaris committed:
fix: ignore string with parenthesis

if it is not an sql function

netchampfaris committed:
Merge branch 'develop' into refactor-qb-engine

netchampfaris committed:
chore: indentation fix

netchampfaris committed:
fix: pass fields explicitly

- to prevent addition of default `name` field
- also, add fields only if it is a select query

netchampfaris committed:
fix: use Field objects as is in apply_filter

ankush commented:
ERPNext: https://github.com/frappe/erpnext/actions/runs/3873879095

netchampfaris committed:
fix: set default order_by direction to desc

netchampfaris committed:
fix: list filter

filters as list must always be list of list

netchampfaris committed:
fix: allow list of dict in filters

netchampfaris committed:
Merge branch 'develop' into refactor-qb-engine

ankush commented:
> I have also changed all instances qb.engine.get_query to qb.get_query because I ran into a problem where the same instance of Engine was modified during a query.

can you share example of this? `Engine` is recycled with every request/BG job. So curious to see where this pollution is happening.

ankush commented:
TODO:
- [ ] Profiling for perf impact.

netchampfaris committed:
fix: refactor

- move operator map in separate file
- remove unnecessary code
- organize functions

netchampfaris committed:
fix: allow Table instance

netchampfaris committed:
fix: move pluck to run

netchampfaris committed:
fix: support list of str or int in filters

netchampfaris committed:
fix: don't cast integer value in filter

netchampfaris committed:
fix: handle empty list for "in" and "not in"

netchampfaris committed:
fix: handle empty string passed to filters

netchampfaris committed:
test: tests for various filter options

netchampfaris committed:
fix: handle empty list as filters

ankush commented on frappe/__init__.py:
```diff
@@ -244,7 +244,7 @@ def init(site: str, sites_path: str = ".", new_site: bool = False) -> None:
 	local.dev_server = _dev_server
 	local.qb = get_query_builder(local.conf.db_type or "mariadb")
-	local.qb.engine = get_qb_engine()
+	local.qb.get_query = get_query
```
Don't see a point in making a static import dynamic. Makes it harder to discover where code lives for no benefit. If it's "internal" let's not attach it to request locals.

ankush replied:
or is this not supposed to be "internal" anymore?

netchampfaris replied:
@ankush 
I want people to use `frappe.qb.get_query` instead of `frappe.db.get_list`. I want it to be "public".
Is there any other way to make it public and not attach it to request locals?

ankush replied:
It's a replacement for `get_all` (sorta, get_list replacement needs perm checks) 

We can leave this here if it's gonna be public.

ankush commented on frappe/database/utils.py:
```diff
@@ -34,6 +34,14 @@ def is_pypika_function_object(field: str) -> bool:
 
+def get_doctype_name(table_name: str) -> str:
+	if "tab" in table_name:
+		table_name = table_name.replace("tab", "")
```
tab can be part of doctype name too. 

![image](https://user-images.githubusercontent.com/9079960/212833932-1e76d6af-d761-44c3-a789-517fef213964.png)

ankush commented on frappe/database/query.py:
```diff
@@ -165,9 +165,7 @@ def _apply_filter(
 				)
 
-		if isinstance(_value, (str, int)):
-			_value = str(_value)
```
This was done to handle primary keys IIRC 

@phot0n

netchampfaris replied:
The generated query is more accurate but the results are the same. Is there anything I am missing?
```py
frappe.qb.get_query("GP Discussion", filters={'name': 1})
SELECT `name` FROM `tabGP Discussion` WHERE `name`=1

frappe.qb.get_query("GP Discussion", filters={'name': "1"})
SELECT `name` FROM `tabGP Discussion` WHERE `name`='1'
```

ankush replied:
Yep. Seems to be compatible with postgres too.

ankush commented on frappe/database/query.py:
```diff
@@ -111,11 +111,14 @@ def apply_filters(
-				elif isinstance(filter, (list, tuple)):
-					self.apply_list_filters(filter)
+			if all(isinstance(d, (str, int)) for d in filters):
+				self.apply_dict_filters({"name": ("in", filters)})
```
Is this sugar really required? Also, it's not supported in other places (get_all) to best of my knowledge, so again more inconsistencies. 

```
frappe.db.get_value("User", ["User1", "User2"])
```

ankush replied:
Also get_all does weird argument gymnastic so list filters will become fields to select 

https://github.com/frappe/frappe/blob/53f81549078e3bdc3fcb5d29109eabc4457fd42c/frappe/model/db_query.py#L122-L124

netchampfaris replied:
> Is this sugar really required?

I don't like it either. But the previous implementation of `frappe.qb.engine.get_query` supported it, so I added it for feature parity. Should I remove it?

netchampfaris replied:
> Also get_all does weird argument gymnastic so list filters will become fields to select

Let's not do this mistake again? `get_all` and `get_query` are different APIs. `get_all` can continue to support it's "weird magic" while using `get_query` under the hood.

ankush commented on frappe/database/query.py:
```diff
@@ -173,6 +173,9 @@ def _apply_filter(
 			_value = self.get_function_object(_value)
 
+		if isinstance(_value, (list, tuple)) and not _value:
+			_value = ("",)
```
We are assuming that the user meant this. Also postgres won't like type-mismatches like these. 

In most cases user wont knowingly pass empty list, the list itself is result of some query or operation that yielded nothing.

netchampfaris replied:
If you pass an empty list, it generates an invalid query. There are legit places in code that depend on this behavior, I don't like it either.

If you think about it, if you pass an empty list, you don't expect any records to match and it returns an empty list, which is what this line does by making it a valid query.

ankush commented on frappe/database/query.py:
```diff
@@ -98,7 +98,7 @@ def apply_filters(
 		self,
 		filters: dict[str, str | int] | str | int | list[list | str | int] | None = None,
 	):
-		if not filters:
```
What does empty string do here? 

`select * where name == ""`

That doesn't make any sense to me. PK is always set in framework.

netchampfaris replied:
This is an older version of this line. It is now: `if filters is None`
https://github.com/frappe/frappe/pull/19405/files#diff-d578653ff2ad9f874d89ebb450b11619cc7771b9ce08dd03de2da25abc29cbc8R101

netchampfaris committed:
fix: only replace "tab" at the beginning
